templates/template_processor.go

```diff
@@ -0,0 +1,135 @@
+		return err
+	}
+
+	if isText {
+		return processTemplate(path, templateFolder, outputFolder, variables)
+	} else {
+		return copyFile(path, templateFolder, outputFolder, variables)
+	}
+}
+
+// Create the given directory, which is in templateFolder, in the given outputFolder
+func createOutputDir(dir string, templateFolder string, outputFolder string) error {
+	destination, err := outPath(dir, templateFolder, outputFolder)
+	if err != nil {
+		return err
+	}
+
+	util.Logger.Printf("Creating folder %s", destination)
+	return os.MkdirAll(destination, 0777)
+}
+
+// Compute the path where the given file, which is in templateFolder, should be copied in outputFolder
+func outPath(file string, templateFolder string, outputFolder string) (string, error) {
+	// TODO process template syntax in paths
+	templateFolderAbsPath, err := filepath.Abs(templateFolder)
+	if err != nil {
+		return "", errors.WithStackTrace(err)
+	}
+
+	fileAbsPath, err := filepath.Abs(file)
+	if err != nil {
+		return "", errors.WithStackTrace(err)
+	}
+
+	relPath, err := filepath.Rel(templateFolderAbsPath, fileAbsPath)
+	if err != nil {
+		return "", errors.WithStackTrace(err)
+	}
+
+	return path.Join(outputFolder, relPath), nil
+}
+
+// Copy the given file, which is in templateFolder, to outputFolder
+func copyFile(file string, templateFolder string, outputFolder string, variables map[string]string) error {
+	destination, err := outPath(file, templateFolder, outputFolder)
+	if err != nil {
+		return err
+	}
+
+	util.Logger.Printf("Copying %s to %s", file, destination)
+	return util.CopyFile(file, destination)
+}
+
+// Run the template at templatePath, which is in templateFolder, through the Go template engine with the given
+// variables as data and write the result to outputFolder
+func processTemplate(templatePath string, templateFolder string, outputFolder string, variables map[string]string) error {
+	destination, err := outPath(templatePath, templateFolder, outputFolder)
+	if err != nil {
+		return err
+	}
+
+	util.Logger.Printf("Processing template %s and writing to %s", templatePath, destination)
+	bytes, err := ioutil.ReadFile(templatePath)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	out, err := renderTemplate(templatePath, string(bytes), variables)
+	if err != nil {
+		return err
+	}
+
+	return util.WriteFileWithSamePermissions(templatePath, destination, []byte(out))
+}
+
+// Return true if this is a path that should not be copied
+func shouldSkipPath(path string, templateFolder string) bool {
+	return path == templateFolder || path == config.BoilerPlateConfigPath(templateFolder)
+}
+
+// Render the template at templatePath, with contents templateContents, using the Go template engine, passing in the
+// given variables as data.
+func renderTemplate(templatePath string, templateContents string, variables map[string]string) (string, error) {
+	tmpl, err := template.New(templatePath).Funcs(CreateTemplateHelpers(templatePath)).Parse(templateContents)
```

Comment:
Most likely because parsing is slow. That doesn't matter if you only use the template once, but if you are using it over and over again, it's nice to do the parsing just once, and get back a reusable object that has an efficient way to render using the already-parsed content.